### PR TITLE
Spec basic instructions

### PR DESCRIPTION
**Pull request by rossberg**
This updates the spec for load, store, rmw, memory.size & .grow instructions.

* Refactor and extend formal rules to allow atomic instructions on unshared memory.
* Complete redo of rmw rules.
* Update prose spec everywhere.
* Various fixes and clean-ups.

### CODE REVIEW

**Comment by rossberg** [comment]
Unfortunately, it looks like MathJax is messing up the notation pretty badly. I tried a few quick hacks, but to no avail. Leaving that problem for later.

**Comment by rossberg** [comment]
@conrad-watt, PTAL.

File: document/core/exec/instructions.rst [hunk 1]
```diff
@@ -381,7 +381,7 @@ Memory Instructions
 ~~~~~~~~~~~~~~~~~~~
 
 .. note::
-   The alignment :math:`\memarg.\ALIGN` in load and store instructions does not affect the semantics.
+   The alignment :math:`\memarg.\ALIGN` in load and store instructions does not affect the semantics of accesses to :ref:`unshared <syntax-unshared>` :ref:`memories <syntax-mem>`.
```
Comment (conrad-watt) [inline document/core/exec/instructions.rst:384]: Even for shared memories, non-atomic accesses are allowed to have incorrect alignment hints, I think?

File: document/core/exec/instructions.rst [hunk 2]
```diff
@@ -390,98 +390,98 @@ Memory Instructions
 
 .. _exec-load:
 .. _exec-loadn:
-.. _exec-atomic-load:
-.. _exec-atomic-loadn:
 
-:math:`t\K{.}\ATOMIC^?.\LOAD~\memarg` and :math:`t\K{.}\ATOMIC^?.\LOAD{N}\K{\_}\sx~\memarg`
-...........................................................................................
+:math:`t\K{.}\LOAD{N}(\K{\_}\sx)^?~\memarg`
```
Comment (conrad-watt) [inline document/core/exec/instructions.rst:394]: I think the _`N`_ should be inside the option bracket?

File: document/core/exec/instructions.rst [hunk 3]
```diff
@@ -618,19 +624,23 @@ Memory Instructions
...
 
-5. Let :math:`\X{mem}` be the :ref:`memory instance <syntax-meminst>` :math:`S.\SMEMS[a]`.
+   a. Let :math:`\X{mem}` be the :ref:`memory instance <syntax-meminst>` :math:`S.\SMEMS[a]`.
 
-6. Let :math:`\X{sz}` be the length of :math:`\X{mem}.\MIDATA` divided by the :ref:`page size <page-size>`.
+   b. Let :math:`n` be the length of :math:`\X{mem}.\MIDATA`.
+
+5. Else:
+
+   a. Perform the :ref:`action <syntax-act>` :math:`(\ARD~a.\LLEN~n)` to read the length :math:`n` of the shared :ref:`memory instance <syntax-meminst>` at :ref:`memory address <syntax-memaddr>` :math:`a`.
```
Comment (conrad-watt) [inline document/core/exec/instructions.rst:641]: The read action here should be atomic/sc.

File: document/core/exec/instructions.rst [hunk 4]
```diff
@@ -666,33 +677,51 @@ Memory Instructions
...
+      iv. Let :math:`\X{sz}` be :math:`l` divided by the :ref:`page size <page-size>`.
 
-10. Either, try :ref:`growing <grow-mem>` :math:`\X{mem}` by :math:`n` :ref:`pages <page-size>`:
+      v. Push the value :math:`\I32.\CONST~\X{sz}` to the stack.
 
-   a. If it succeeds, push the value :math:`\I32.\CONST~\X{sz}` to the stack.
+   b. Or:
 
-   b. Else, push the value :math:`\I32.\CONST~\X{err}` to the stack.
+      i. Perform the :ref:`action <syntax-act>` :math:`(\ARD~a.\LLEN~l)` to read the length :math:`l` of the shared :ref:`memory instance <syntax-meminst>` at :ref:`memory address <syntax-memaddr>` :math:`a`.
```
Comment (conrad-watt) [inline document/core/exec/instructions.rst:722]: The read here should be atomic/sc.

File: document/core/exec/instructions.rst [hunk 5]
```diff
@@ -737,243 +768,354 @@ Memory Instructions
...
+    a. Perform the :ref:`action <syntax-act>` :math:`(\ARD~a.\LLEN~n)` to read the length :math:`n` of the shared :ref:`memory instance <syntax-meminst>` at :ref:`memory address <syntax-memaddr>` :math:`a`.
 
-17. If :math:`N` is part of the instruction, then:
+    b. If :math:`n` is smaller than :math:`\X{ea} + N/8`, or :math:`\X{ea}` is not divisible by :math:`N/8`, then:
 
-    a. Let :math:`n` be the result of computing :math:`\wrap_{|t|,N}(m)`.
+       i. Trap.
 
-    b. Let :math:`b^\ast_w` be the byte sequence :math:`\bytes_{\iN}(n)`.
+    c. Perform the atomic :ref:`action <syntax-act>` :math:`(\ARMW_{\ord}~a.\LDATA[\X{ea}]~b_{\F{r}}^\ast~b_{\F{r}}^\ast)` to read the bytes :math:`b_{\F{r}}^\ast` from data offset :math:`\X{ea}` of the shared :ref:`memory instance <syntax-meminst>` at :ref:`memory address <syntax-memaddr>` :math:`a` and replace them with bytes :math:`b_{\F{w}}^\ast` (which are defined below).
```
Comment (conrad-watt) [inline document/core/exec/instructions.rst:863]: The _`ord`_ subscript is never defined (should just be atomic/sc).

Comment (conrad-watt) [reply document/core/exec/instructions.rst:863]: Also, `b_r*` seems to be repeated twice when the second component should be `b_w*` (copy-paste nit).

File: document/core/exec/instructions.rst [hunk 6]
```diff
@@ -737,243 +768,354 @@ Memory Instructions
...
-    a. Trap.
+    a. Perform the :ref:`action <syntax-act>` :math:`(\ARD~a.\LLEN~n)` to read the length :math:`n` of the shared :ref:`memory instance <syntax-meminst>` at :ref:`memory address <syntax-memaddr>` :math:`a`.
+
+    b. If :math:`n` is smaller than :math:`\X{ea} + N/8`, or :math:`\X{ea}` is not divisible by :math:`N/8`, then:
 
-16. Let :math:`b^\ast_r` be the byte sequence :math:`\X{mem}.\MIDATA[\X{ea} \slice N/8]`.
+       i. Trap.
 
-17. Let :math:`c_1` be the integer for which :math:`bytes_{\iN}(m) = b^\ast_r`.
+    c. Perform the atomic :ref:`action <syntax-act>` :math:`(\ARMW_{\ord}~a.\LDATA[\X{ea}]~b_{\F{r}}^\ast~b_{\F{r}}^\ast)` to read the bytes :math:`b_{\F{r}}^\ast` from data offset :math:`\X{ea}` of the shared :ref:`memory instance <syntax-meminst>` at :ref:`memory address <syntax-memaddr>` :math:`a` and replace them with bytes :math:`b_{\F{w}}^\ast` (which are defined below).
```
Comment (conrad-watt) [inline document/core/exec/instructions.rst:1006]: rmw copy-paste nit replicated here (although as described above this section needs to change).

File: document/core/exec/instructions.rst [hunk 7]
```diff
@@ -737,243 +768,354 @@ Memory Instructions
...
 
 
-.. _exec-atomic-rmw-cmpxchg:
-.. _exec-atomic-rmwn-cmpxchg:
+.. _exec-atomic.rmw.cmpxchg:
+.. _exec-atomic.rmwn.cmpxchg:
 
-:math:`t\K{.}\ATOMICRMW{.}\ATOMICCMPXCHG~\memarg` and :math:`t\K{.}\ATOMICRMW{N}\K{\_u.}\ATOMICCMPXCHG~\memarg`
-...............................................................................................................
+:math:`t\K{.}\ATOMICRMW({N}\K{\_u})^?\K{.}\ATCMPXCHG~\memarg`
```
Comment (conrad-watt) [inline document/core/exec/instructions.rst:961]: High-level:

cmpxchg should only perform an rmw action if the comparison succeeds. This means that at the operational level, the instruction non-deterministically either emits an atomic read action where the observed value is != the argument, or an rmw action where the read value is == the argument.

File: document/core/exec/instructions.rst [hunk 8]
```diff
@@ -390,98 +390,98 @@ Memory Instructions
...
 
-    b. Let :math:`c` be the result of computing :math:`\extend\F{\_}\sx_{N,|t|}(n)`.
+    b. If :math:`n` is smaller than :math:`\X{ea} + N/8`, or both :math:`\ord` is :math:`\SEQCST` and :math:`\X{ea}` is not divisible by :math:`N/8`, then:
 
-13. Else:
+       i. Trap.
+
+    c. Perform the :ref:`action <syntax-act>` :math:`(\ARD_{\ord}~a.\LDATA[\X{ea}]~b^\ast)` to read the bytes :math:`b^\ast` from data offset :math:`\X{ea}` of the shared :ref:`memory instance <syntax-meminst>` at :ref:`memory address <syntax-memaddr>` :math:`a`.
+
+11. If :math:`N` and :math:`\sx` are part of the instruction, then:
```
Comment (conrad-watt) [inline document/core/exec/instructions.rst:435]: can conditionals like this be factored more like the formal semantics?

i.e. have a few preamble lines that state
"1. N = |t| if N is not part of the instruction"

some of the RMWs especially have a lot of case-splitting in the text on this

File: document/core/exec/instructions.rst [hunk 9]
```diff
@@ -666,33 +677,51 @@ Memory Instructions
...
+
+      ii. Else, push the value :math:`\I32.\CONST~\X{err}` to the stack.
+
+   d. Or, push the value :math:`\I32.\CONST~\X{err}` to the stack.
 
-4. Assert: due to :ref:`validation <valid-memory.grow>`, :math:`S.\SMEMS[a]` exists.
+8. Else:
 
-5. Let :math:`\X{mem}` be the :ref:`memory instance <syntax-meminst>` :math:`S.\SMEMS[a]`.
+   a. Either:
```
Comment (conrad-watt) [inline document/core/exec/instructions.rst:708]: the "try growing memory" text in the non-shared case above was always informal language. Could some of that flavour be added to this case too? i.e.

```
a. Either:
  memory growth succeeds.
  ... let bla bla bla
b. Or:
  memory growth fails.
  ... bla bla bla
```

**Review by conrad-watt — CHANGES_REQUESTED** [review]
Main issue currently is the behaviour of `cmpxchg`.

After this PR, I believe the next steps are to add `fence` and flesh out the semantics of `wait/notify`?

File: document/core/exec/instructions.rst (hunk 7 above)
Comment (conrad-watt) [reply document/core/exec/instructions.rst:961]: Huh, the JS spec also models compareExchange as an rmw even if the comparison fails. I'll double check, but I'm pretty sure that's not correct.

Comment (conrad-watt) [reply document/core/exec/instructions.rst:961]: C++11 models it as just a load in the failure case, so I think this is also a JS spec bug.

File: document/core/valid/instructions.rst [hunk 10]
```diff
@@ -431,228 +431,212 @@ Memory Instructions
...
 
 .. math::
    \frac{
      C.\CMEMS[0] = \memtype
      \qquad
-     2^{\memarg.\ALIGN} = |t|/8
+     2^{\memarg.\ALIGN} = 4
    }{
-     C \vdash t\K{.}\ATOMICRMW\K{.}\ATOMICCMPXCHG~\memarg : [\I32~t~t] \to [t]
+     C \vdash \MEMORYATOMICNOTIFY~\memarg : [\I32~\I64] \to [\I64]
```
[the remark this answers is not in the document]
Comment (rossberg) [reply document/core/valid/instructions.rst]: Yes, the current reduction rules also match that, fixed.

However, the reduction rule for wait seem to have the operand order reversed relative to both validation and overview. Added a todo.

**Review by rossberg — COMMENTED** [review]
Thanks for the thorough review! Yes, I think the next steps will be wait/notify/fence. Do you want to proceed with those?

**Review by conrad-watt — COMMENTED** [review]
Just a few nits.

I can look at fence/wait/notify in the next couple of weeks, but I really need to submit my thesis so I'm finding it hard to manage my time.

File: document/core/exec/instructions.rst [hunk 11]
```diff
@@ -390,98 +390,98 @@ Memory Instructions
...
-    a. Let :math:`n` be the integer for which :math:`\bytes_{\iN}(n) = b^\ast`.
+      i. Trap.
 
-    b. Let :math:`c` be the result of computing :math:`\extend\F{\_}\sx_{N,|t|}(n)`.
+   c. Let :math:`b^\ast` be the byte sequence :math:`\X{mem}.\MIDATA[\X{ea} \slice N/8]`.
 
-13. Else:
+11. Else:
+
+    a. Perform the :ref:`action <syntax-act>` :math:`(\ARD~a.\LLEN~n)` to read the length :math:`n` of the shared :ref:`memory instance <syntax-meminst>` at :ref:`memory address <syntax-memaddr>` :math:`a`.
```
[the remark this answers is not in the document]
Comment (conrad-watt) [reply document/core/exec/instructions.rst:433]: sorry, this comment should have been at line 439 just below

File: document/core/exec/instructions.rst [hunk 12]
```diff
@@ -390,98 +390,98 @@ Memory Instructions
...
+...........................................
 
-.. todo:: update text to match formalism
+1. Let :math:`\ord` be :math:`\UNORD`.
 
-1. Let :math:`F` be the :ref:`current <exec-notation-textual>` :ref:`frame <syntax-frame>`.
+2. If :math:`N` and :math:`\sx` are part of the instruction, then:
 
-2. Assert: due to :ref:`validation <valid-loadn>`, :math:`F.\AMODULE.\MIMEMS[0]` exists.
+   a. Let :math:`\X{st}` be the :ref:`pro forma type <aux-bytes>` :math:`\iN`.
```
[the remark this answers is not in the document]
Comment (rossberg) [reply document/core/exec/instructions.rst]: Updated.

File: document/core/exec/instructions.rst (hunk 11 above)
[the remark this answers is not in the document]
Comment (rossberg) [reply document/core/exec/instructions.rst:433]: Made explicit in the prose. In the formal semantics that's implicit by the side conditions, particularly, its use as an argument to bytes^-1.

File: document/core/exec/instructions.rst [hunk 13]
```diff
@@ -502,86 +505,88 @@ Memory Instructions
...
+.................................
 
-.. todo:: update text to match formalism
+1. Let :math:`\ord` be :math:`\UNORD`.
 
-1. Let :math:`F` be the :ref:`current <exec-notation-textual>` :ref:`frame <syntax-frame>`.
+2. If :math:`N` is part of the instruction, then:
 
-2. Assert: due to :ref:`validation <valid-storen>`, :math:`F.\AMODULE.\MIMEMS[0]` exists.
+   a. Let :math:`\X{st}` be the :ref:`pro forma type <aux-bytes>` :math:`\iN`.
```
[the remark this answers is not in the document]
Comment (rossberg) [reply document/core/exec/instructions.rst]: Fixed.

File: document/core/exec/instructions.rst [hunk 14]
```diff
@@ -714,266 +745,372 @@ Memory Instructions
...
 
-.. todo:: update text to match formalism
+:math:`t\K{.}\ATOMICRMW({N}\K{\_u})^?\K{.}\atop~\memarg`
+........................................................
 
-1. Let :math:`F` be the :ref:`current <exec-notation-textual>` :ref:`frame <syntax-frame>`.
+1. If :math:`N` and :math:`\K{\_u}` are part of the instruction, then:
 
-2. Assert: due to :ref:`validation <valid-atomic-rmwn>`, :math:`F.\AMODULE.\MIMEMS[0]` exists.
+   a. Let :math:`\X{st}` be the :ref:`pro forma type <aux-bytes>` :math:`\iN`.
```
[the remark this answers is not in the document]
Comment (rossberg) [reply document/core/exec/instructions.rst]: Fixed.

File: document/core/exec/instructions.rst [hunk 15]
```diff
@@ -714,266 +745,372 @@ Memory Instructions
...
+    a. Perform the :ref:`action <syntax-act>` :math:`(\ARD~a.\LLEN~n)` to read the length :math:`n` of the shared :ref:`memory instance <syntax-meminst>` at :ref:`memory address <syntax-memaddr>` :math:`a`.
 
-18. Else:
+    b. If :math:`n` is smaller than :math:`\X{ea} + N/8`, or :math:`\X{ea}` is not divisible by :math:`N/8`, then:
 
-    a. Let :math:`b^\ast_w` be the byte sequence :math:`\bytes_t(m)`.
+       i. Trap.
 
-19. Replace the bytes :math:`\X{mem}.\MIDATA[\X{ea} \slice N/8]` with :math:`b^\ast_w`.
+    c. Perform the atomic :ref:`action <syntax-act>` :math:`(\ARMW_{\SEQCST}~a.\LDATA[\X{ea}]~b_{\F{r}}^\ast~b_{\F{w}}^\ast)` to read the bytes :math:`b_{\F{r}}^\ast` from data offset :math:`\X{ea}` of the shared :ref:`memory instance <syntax-meminst>` at :ref:`memory address <syntax-memaddr>` :math:`a` and replace them with bytes :math:`b_{\F{w}}^\ast` (which are defined below).
```
[the remark this answers is not in the document]
Comment (rossberg) [reply document/core/exec/instructions.rst]: Fixed as above. (Yes, for b*_w this follows from the definition of bytes_st.)

File: document/core/exec/instructions.rst [hunk 16]
```diff
@@ -714,266 +745,372 @@ Memory Instructions
...
+     &\stepto^{(\ARD~a.\LLEN~n)~(\ARMW~a.\LDATA[\X{ea}]~b_{\F{r}}^\ast~b_{\F{w}}^\ast)}&
+     S; F; (t.\CONST~c_1)
+   \end{array}
+   \\ \qquad
+     \begin{array}[t]{@{}r@{~}l@{}}
+     (\iff & S.\SMEMS[a] ~\mbox{undefined} \\
+     \wedge & \X{ea} + N/8 \leq n \\
+     \wedge & \X{ea} \mod N/8 = 0 \\
+     \wedge & c_1 = \extendtu_{\X{st},t}(\bytes^{-1}_{\X{st}}(b_{\F{r}}^\ast)) \\
+     \wedge & b_{\F{r}}^\ast = \bytes_{\X{st}}(\wrapt_{t,\X{st}}(\atop_t(c_1, c_2)))) \\
```
[the remark this answers is not in the document]
Comment (rossberg) [reply document/core/exec/instructions.rst]: Fixed.

File: document/core/exec/instructions.rst [hunk 17]
```diff
@@ -714,266 +745,372 @@ Memory Instructions
...
+.............................................................
 
-.. todo:: update text to match formalism
+.. todo:: should only emit |ARMW| if the condition is true?
 
-1. Let :math:`F` be the :ref:`current <exec-notation-textual>` :ref:`frame <syntax-frame>`.
+1. If :math:`N` and :math:`\K{\_u}` are part of the instruction, then:
 
-2. Assert: due to :ref:`validation <valid-atomic-rmwn-cmpxchg>`, :math:`F.\AMODULE.\MIMEMS[0]` exists.
+   a. Let :math:`\X{st}` be the :ref:`pro forma type <aux-bytes>` :math:`\iN`.
```
[the remark this answers is not in the document]
Comment (rossberg) [reply document/core/exec/instructions.rst]: Fixed.

File: document/core/exec/instructions.rst [hunk 18]
```diff
@@ -714,266 +745,372 @@ Memory Instructions
...
-    a. Let :math:`n` be the result of computing :math:`\wrap_{|t|,N}(c_3)`.
+14. Else:
+
+    a. Perform the :ref:`action <syntax-act>` :math:`(\ARD~a.\LLEN~n)` to read the length :math:`n` of the shared :ref:`memory instance <syntax-meminst>` at :ref:`memory address <syntax-memaddr>` :math:`a`.
+
+    b. If :math:`n` is smaller than :math:`\X{ea} + N/8`, or :math:`\X{ea}` is not divisible by :math:`N/8`, then:
+
+       i. Trap.
+
+    c. Perform the atomic :ref:`action <syntax-act>` :math:`(\ARMW_{\SEQCST}~a.\LDATA[\X{ea}]~b_{\F{r}}^\ast~b_{\F{w}}^\ast)` to read the bytes :math:`b_{\F{r}}^\ast` from data offset :math:`\X{ea}` of the shared :ref:`memory instance <syntax-meminst>` at :ref:`memory address <syntax-memaddr>` :math:`a` and replace them with bytes :math:`b_{\F{w}}^\ast` (which are defined below).
```
[the remark this answers is not in the document]
Comment (rossberg) [reply document/core/exec/instructions.rst]: Fixed.

**Review by rossberg — COMMENTED** [review]
Thanks! No stress, I thought your thesis was already due in September.

**Comment by conrad-watt** [comment]
> Thanks! No stress, I thought your thesis was already due in September.

yes

**Review by conrad-watt — APPROVED** [review]
LGTM

**Pull request merged**